### PR TITLE
feat: init kp

### PR DESCRIPTION
**Pull request by bludot**
This change is [Reviewable](https://reviewable.io/reviews/honestbank/kp/1)

### CODE REVIEW

File: kp/config.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package kp
+
+type Config struct {
+	KafkaConfig KafkaConfig `json:"kafka"`
+}
+
+type KafkaConfig struct {
+	KafkaBootstrap string `env:"CONFIG__KAFKA__BOOTSTRAP" default:"localhost:9092" json:"kafka_bootstrap"`
```
Comment (cyberhck) [inline kp/config.go:8]: isn't this a library like? I don't think it should read config like this, I think it should have asked for connection instead, no?

File: kp/consumer.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+func (consumer *ConsumerStruct) ProcessMessage(message *sarama.ConsumerMessage) error {
+	unmarshaledMessage, retries, err := UnmarshalStringMessage(string(message.Value))
+	if err != nil {
+		log.Printf("Error unmarshaling message: %v", err)
+
+		return err
+	}
+	if retries >= consumer.retries {
+		log.Printf("Message has exceeded retries, sending to dead letter topic")
```
Comment (cyberhck) [inline kp/consumer.go:64]: are you sending to dead-letter topic though?

File: kp/kp.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+		log.Panicf("Error closing client: %v", err)
+	}
+}
+
+func (k *KP) Stop() {
+	log.Println("Stopping Kafka consumer")
+	k.client.Close()
+}
+
+func toggleConsumptionFlow(client sarama.ConsumerGroup, isPaused *bool) {
```
Comment (cyberhck) [inline kp/kp.go:112]: what's this for and why would we use it?

**Review by cyberhck — CHANGES_REQUESTED** [review]

File: kp/kp.go (hunk 3 above)
Comment (bludot) [reply kp/kp.go:112]: you can pause jobs using sigusr1, this handles it

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<!--
+# Pull Request Instructions
+
+* All PRs should reference an issue in our issue tracker. If one doesn't exist, please create one!
+* PR titles should follow https://www.conventionalcommits.org.
```
Comment (nithipanhn) [inline .github/pull_request_template.md:5]: Please use the new template

File: .idea/.gitignore [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
```
Comment (nithipanhn) [inline .idea/.gitignore:1]: add .idea to gitignore

File: kp/consumer.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	}
+	if retries >= consumer.retries {
+		log.Printf("Message has exceeded retries, removing from kafka")
+
+		return nil
+	}
+	err = consumer.Processor(string(message.Key), unmarshaledMessage, retries, message)
+	if err != nil {
+		marshaledMessage := MarshalStringMessage(unmarshaledMessage, retries+1)
+		err = consumer.producer.ProduceMessage(consumer.deadLetterTopic, string(message.Key), marshaledMessage)
```
Comment (nithipanhn) [inline kp/consumer.go]: I'm trying to understand. If the processor returns an error, it reproduces the message into the dead letter topic with retires?

File: config.go [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package kp
+
+type Config struct {
+	KafkaConfig KafkaConfig `json:"kafka"`
+}
+
+type KafkaConfig struct {
+	KafkaBootstrapServers string `json:"kafka_bootstrap"`
```
[the remark this answers is not in the document]
Comment (cyberhck) [reply config.go:8]: I'd rather ask for array of string, and pass it down to client.

File: kp/consumer.go (hunk 6 above)
Comment (nithipanhn) [reply kp/consumer.go]: IMO, the flow should be 

                        
Screen Shot 2565-06-27 at 15 27 41

Comment (nithipanhn) [reply kp/consumer.go]: or

Comment (nithipanhn) [reply kp/consumer.go]: Screen Shot 2565-06-27 at 15 31 47

Comment (bludot) [reply kp/consumer.go]: its like the second one, but im not listening to deal letter

**Review by cyberhck — APPROVED** [review]

**Review by nithipanhn — APPROVED** [review]

**Pull request merged**